### PR TITLE
LibJS: Initial factoring towards a standalone LibGC

### PR DESCRIPTION
**Pull request by shannonbooth**
These are the initial non-significant changes towards a LibGC.

Next steps to get this fully working:

1. Move LibJS's Heap folder into a LibGC
2. Smuggle in the `JS::VM` pointer to the `GC::Heap` as an opaque `void*`
4. Create a `JS::Cell` subclass from `GC::Cell` which contains the `void initialize(JS::Realm&)` virtual function
5. `JS::Cell` also provides the `VM& vm()` getter from casting it's `void*` private data
6. Create a `JS::Heap` subclass from `GC::Heap` that provides the `allocate<T>` function (i.e is aware of `JS::Realm&` initialization)
7. Create a `GC::NanBoxedValue` which is aware of tagging a `NanBoxedValue` as a `GC::Cell`. `JS::Value` inherits from `GC::NanBoxedValue`. visits to a `JS::Value` are through `visit_edges(NanBoxedValue const&)`
8. ...Rename a bazzilion places to the new namespace, likely also renaming these classes to the new preferred names (such as `GC::Ptr` and `GC::Ref`

Ideas for a better approach than those steps welcome. All of those above steps are somewhat intertangled, so may need to be one ginormous commit.

### CODE REVIEW

**Comment by ADKaster** [comment]
Without looking at the code yet, I have a  question:

What does this refactor do to the size of JS::Cell? The minimum size previously was I think 24 bytes? I think we fixed the issue with FreeListEntry being that size a while back, but the heap poisoning implementation is pretty sensitive to the minimum cell size.

**Comment by shannonbooth** [comment]
This MR is mostly just shuffling stuff around since what's described in the description unforuntately has to be done in one big chunk altogether as far as I can tell to remain atomic. So I thought I may as well get the stuff that can be done separately out of the way

**Comment by shannonbooth** [comment]
oh duh, the virtual would increase size, let me check properly

**Comment by shannonbooth** [comment]
Okay, yeah 24 bytes before and after from a bit of a hacky test

File: Userland/Libraries/LibJS/Heap/Heap.cpp [hunk 1]
```diff
@@ -266,7 +267,9 @@ void Heap::collect_garbage(CollectionType collection_type, bool print_report)
 
 void Heap::gather_roots(HashMap<Cell*, HeapRoot>& roots)
 {
-    vm().gather_roots(roots);
+    if (m_gather_roots)
+        m_gather_roots(roots);
```
Comment (ADKaster) [inline Userland/Libraries/LibJS/Heap/Heap.cpp:271]: This name is a bit confusing when the method name is also `gather_roots`. Perhaps, "m_gather_embedder_roots" or similar?

File: Userland/Libraries/LibWeb/WebIDL/Promise.h [hunk 2]
```diff
@@ -10,13 +10,11 @@
 #include <LibJS/Forward.h>
 #include <LibJS/Runtime/PromiseCapability.h>
 #include <LibJS/Runtime/Value.h>
-#include <LibJS/SafeFunction.h>
 #include <LibWeb/Forward.h>
 #include <LibWeb/WebIDL/ExceptionOr.h>
 
 namespace Web::WebIDL {
 
-// NOTE: This is Function, not SafeFunction, because they get stored in a NativeFunction anyway, which will protect captures.
```
[the remark this answers is not in the document]
Comment (shannonbooth) [reply Userland/Libraries/LibWeb/WebIDL/Promise.h:19]: I don't think so, it's leftover cruft from refactoring - The original change adding this was: https://github.com/LadybirdBrowser/ladybird/commit/c10f9856a49be458f8568ef2e7e1e531ddb7e550

But the comment isn't relevant any more after https://github.com/LadybirdBrowser/ladybird/commit/ffac32d20e6e47381df6c73003cc32ede2ebf19b

**Review by ADKaster — APPROVED** [review]

**Comment by shannonbooth** [comment]
Finally got around to resolving the conflict on this

**Pull request closed without merging**